### DISCUSSION TITLE
Studio Activity ForEach sending out a list of emails for approvals

### DISCUSSION THREAD

**Discussion by 7studios** [elsa-workflows/elsa-studio#227] [General]
I am trying to create a workflow similar to the old document workflow in Elsa Studio. I'm facing issues with using the ForEach activity to send multiple emails from a JSON payload like the one below. Could anyone share an example that addresses this?

{
  "Approvers": [
    {
      "id": 1,
      "email": "bob@acme.com",
      "name": "Bob Jones"
    },
    {
      "id": 2,
      "email": "joe@acme.com",
      "name": "Joe Smith"
    }
  ]
}

**Comment by glime-ai** [comment]
To create a workflow in Elsa Studio that sends multiple emails using the ForEach activity with a JSON payload like the one you provided, you can follow these steps:

1. **Define the JSON Payload**: Store the JSON payload as a workflow variable or input it directly into the ForEach activity if it's coming from an external trigger.

2. **Add a ForEach Activity**: Drag and drop the ForEach activity into your workflow designer.

3. **Configure the ForEach Activity**: Set the `CollectionExpression` property of the ForEach activity to the JSON array you want to iterate over. If you stored the JSON payload as a workflow variable named `ApproversList`, you would use a Liquid expression like `{{ ApproversList }}`.

4. **Add a Send Email Activity Inside the ForEach Loop**: Inside the ForEach activity, add a Send Email activity. Configure the Send Email activity to use the current item's email address. You can access the current item's email using a Liquid expression like `{{ CurrentItem.email }}`.

5. **Handle the Loop Execution**: Ensure that the ForEach activity is connected correctly to the Send Email activity and that the workflow logic continues after the ForEach loop is completed.

Here's a simplified example of how the JSON might be used in a ForEach loop to send emails:

```json
{
  "activities": [
    {
      "id": "foreach",
      "type": "ForEach",
      "properties": {
        "collectionExpression": "{{ ApproversList }}"
      }
    },
    {
      "id": "send-email",
      "type": "SendEmail",
      "properties": {
        "to": "{{ CurrentItem.email }}",
        "subject": "Approval Needed",
        "body": "Hello {{ CurrentItem.name }}, please review the document."
      }
    }
  ],
  "connections": [
    {
      "sourceActivityId": "foreach",
      "targetActivityId": "send-email",
      "outcome": "Iterate"
    }
  ],
  "variables": {
    "ApproversList": [
      {
        "id": 1,
        "email": "bob@acme.com",
        "name": "Bob Jones"
      },
      {
        "id": 2,
        "email": "joe@acme.com",
        "name": "Joe Smith"
      }
    ]
  }
}
```

Please note that the actual configuration might differ based on the version of Elsa you are using and the specific details of your workflow. If you encounter any issues, such as the ForEach activity's input dropdown being empty, you can try using a Liquid expression as a workaround, as mentioned in one of the sources.

If you need further assistance or a more detailed example, please let me know.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by 7studios** [reply]
I'm looking for 3.1 activity support.
